FIX: Migrate Mono APIs to CoreCLR-compatible APIs

MiroBrodlovaUnityGit opened the pull request:
## Purpose of this PR

This PR replaces all instances of analyzer-highlighted `Mono` API calls with `CoreCLR`-compatible API calls in com.unity.collections, and is part of Epic [SCP-1555](https://jira.unity3d.com/browse/SCP-1555).

JIRA: https://jira.unity3d.com/browse/SCP-1555

### Further Context

- Unity is steadily moving away from `Mono` to using `CoreCLR`. 
- With this we need to replace all `Mono` API calls with `CoreCLR`-compatible ones. (here's the [list of APIs being replaced](https://docs.google.com/spreadsheets/d/1YRRfjPShJgaaToFZqFpqcE1Nnj6-BbjyeI5rKVc-omM/edit?gid=1326171865#gid=1326171865), and [here's the list](https://docs.google.com/spreadsheets/d/1YRRfjPShJgaaToFZqFpqcE1Nnj6-BbjyeI5rKVc-omM/edit?gid=0#gid=0) of packages that will get updated).
- Epic [SCP-1555](https://jira.unity3d.com/browse/SCP-1555) tracks the work of replacing all `Mono` API calls across all Unity supported packages with `CoreCLR`-compatible ones.
- For more information, or if you have _any_ questions, please visit [#code-reload-safe-api-packages-migration](https://unity.enterprise.slack.com/archives/C09N2RM0L4D) and we'll be happy to help. 🙂

**Note:** More PRs dealing with [SCP-1555](https://jira.unity3d.com/browse/SCP-1555) may come in the future. In some cases we cannot replace all `Mono` API calls with `CoreCLR` API calls yet, as some `CoreCLR` APIs are still being worked on/are not yet public. In these cases we suppress warnings for those particular APIs for now.

## Release Notes

None

## Functional Testing status

Running the default job-suite that is run on new PRs. Please do share if there's some steps I've missed that I must do to properly setup the PR and run your CI jobs!

## Performance Testing Status
Haven't run any additional performance-tests as this PR should just replace Mono API with equivalent CoreCLR API

## Overall Product Risks
Low risk as it just changes Mono API to equivalent CoreCLR API

## Comments to Reviewers

If there are any jobs/tests we should run to make sure there are no ripple effects do share and we will kick them off! :) And, if you know of any locations we may have missed updating or any details that we should know about, please do share!

MiroBrodlovaUnityGit committed:
Migrate AppDomain.GetAssemblies() to CurrentAssemblies.GetLoadedAssemblies()

- LinkFileGenerator.cs: Add conditional compilation (line 43)
- Added using UnityEngine.Assemblies with conditional directive
- Maintains backward compatibility with older Unity versions

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=2289) <br/>All committers have signed the CLA.

MiroBrodlovaUnityGit committed:
Migrate additional AppDomain.GetAssemblies() usages to CoreCLR-compatible APIs

Add conditional compilation to replace AppDomain.CurrentDomain.GetAssemblies()
with CurrentAssemblies.GetLoadedAssemblies() in:
- InputManager.cs
- InputParameterEditor.cs
- CoreTests_Controls.cs
- InputTestFixture.cs

Maintains backward compatibility with Unity versions prior to 6000.5.

MiroBrodlovaUnityGit commented:
Just double checking if there's anything I should do/run/add for this to become merge:able? :) 

I also see several jobs failing, but my suspicion it's not related to these changes maybe? I checked the history of some of the failing jobs on other branches and seems they were failing on other branches too not too long ago? But maybe I'm misunderstanding!

Appreciate any input and help! :)

jfreire-unity committed:
Merge branch 'develop' into fix/remove-mono-api

ekcoh reviewed (approved):
Thanks for doing this. Looks good.

ekcoh commented:
@MiroBrodlovaUnityGit I think CI failure was unrelated, I started a rerun. Also @jfreire-unity added our QA @Pauliusd01 to the PR in case he would like to check something. Otherwise I think this can land, so feel free to merge @Pauliusd01 once you have approved.

jfreire-unity commented:
There was some CI issues that were fixed and I updated this PR with latest `develop` branch changes.
Also, I just want to make sure if @Pauliusd01 spots something we didn't. Otherwise looks good to be merged

Pauliusd01 reviewed (approved):
LGTM, tried running a couple of samples on 22.3 (mono player/playmode) and latest trunk (mono only player/playmode), did not notice any issues. I wasn't able to build the coreclr version editor on trunk though, kept erroring out. If there's a specific way I need to do it or I need a specific revision then let me know on slack and I'll check that one as well

MiroBrodlovaUnityGit commented:
Thanks all for your help on this/reviewing and checking it so thoroughly! :) 

>  I wasn't able to build the coreclr version editor on trunk though, kept erroring out.

Hoping this turned out OK! Let me know if there's anything I should/could do :)

ekcoh committed:
Merge branch 'develop' into fix/remove-mono-api

ekcoh committed:
Merge branch 'develop' into fix/remove-mono-api

ekcoh committed:
Merge branch 'develop' into fix/remove-mono-api
